journal/202111.md

```diff
@@ -0,0 +1,525 @@
+
+The `uint256` package became the first step of a [long-term plan](https://github.com/decred/dcrd/issues/2786) to develop a new `primitives` module. It will become a new fundamental building block of consensus code by consolidating key concepts such as transactions, block headers, blocks, and sequence lock definitions, which are currently spread across multiple packages. Refined abstractions will make the distinction between consensus primitives and the data sharing protocol for greater clarity, and allow for significant performance optimizations.
+
+Progress of the `stdscript` package (an effort to decouple "standard" from "consensus" script handling, in order to avoid unintended changes to consensus):
+
+- added [provably pruneable](https://github.com/decred/dcrd/pull/2803) script support. This is a special form of public key script that cannot be spent and is safe to prune from the set of unspent transaction outputs (UTXOs) to save memory. They are primarily useful for anchoring data into the blockchain, as used by [timestamping](https://timestamp.decred.org/) and [Politeia](https://proposals.decred.org/).
+- added support for determining the number [signatures](https://github.com/decred/dcrd/pull/2805) required by a script
+- added extraction of [addresses](https://github.com/decred/dcrd/pull/2807) from standard recognized forms of public key scripts
+- package `stdscript` [moved](https://github.com/decred/dcrd/pull/2810) from the internal staging area to the `txscript` module and is now ready for consumption
+- all packages [converted](https://github.com/decred/dcrd/pull/2808) to use the new `stdscript` package. To make review easier, the change was split into 50 small commits such that everything builds and passes all tests each step of the way.
+- added methods to universally [extract](https://github.com/decred/dcrd/pull/2816) public key hashes and script hashes from stake-related scripts, which helps many applications that don't care about the specific script type (ticket purchase, vote reward, revocation, change, etc)
+
+Other merged changes:
+
+- fill and check [fraud proof](https://github.com/decred/dcrd/pull/2804) data for transactions entering the mempool. This data includes the input amount, block height, and an index in that block. { optionally add 1-2 sentences explaining what "fraud proof" can do and for whom (light clients?), or just a good link }
```

Comment:
It allows the fees to be calculated without also having access to all of the previous outputs the transaction is spending as well as easily locating previous outputs without requiring a full transaction hash index.

Reply:
Oh, so having input amounts "at hand" allows to subtract output amounts and get the fee that will go to the miner, without possibly slow access to previous outputs. Correct?

This is great for performance but I wonder for what use case it got "fraud" in the title?

Reply:
Yes, regarding the first question, it also allows that, which is an additional nice aspect, and something that is highly useful and used in practice, and indeed why I mentioned it, but the fraud part proof part is also implicit in that along with the rest of what I said above.  I guess that isn't super obvious though since you're asking.

Essentially, a fraud proof is a way to prove that data is invalid without requiring access to the entire state of a system.

For example, consider a bad actor attempting to pass off a fraudulent transaction that spends more than the inputs provide.  You can very easily prove the transaction is fraudulent via the fraud proof alone, without also having access to all of the previous outputs the transaction is spending, because it allows you to calculate the total amount of the inputs to compare against the total amount of outputs in order to determine if the total output amount exceeds the total input amount.  In other words, the act of calculating the "fees" also allows fraud detection (and proving) because the "fees" will be negative if the total output amount exceeds the total input amount.

Similarly, if a fraudulent transaction is attempting to pass off an invalid input amount prior to the fraud proof itself being checked, the other fields of the fraud proof allow the prover to quickly locate the very specific output(s) being referenced "without requiring a full transaction hash index" (aka without requiring access to the entire state of the system) in order to prove the fraud.

Reply:
Added a bit of that for education.

That's was very informative and too good to drown in code reviews so I saved it for the future. Thanks.